togglebyte/anathema
# magic1

- lots and lots of changes
- rename and cleanup
- remove unused functions
- vstack & hstack
- zstack
- position and spacer
- spacers are spacing!
- expand widgets, bug fix
- added offset to values
- alignment
- viewport is implemented but needs cleanup
- cleanup the viewport code
- BUGFIX: substract with overflow
- BUGFIX: accidentally removed an offset
- Viewports can now scroll in each direction along each axis. Needs cleanup
- distinguish between inner and outer size, as inner size takes padding into consideration
- added alignment widget
- more readable widget tests
- ignore files
- more tests. found bug with text layout, not fixed yet
- removed what seemed like a mistake
- tests for alignment
- more tests
- more tests
- more tests
- more tests for expand
- clean up tests
- tests for spacers
- fixed failing tests
- fixed failing tests
- fixed failing tests
- renamed For to Loop
- github actions
- code formatting
- CI
- remove unused widgets (for now, can add them back later if they are wanted)
- use 'many' layout for both the vstack and hstack with defaults
- cleaning up warnings
- cleaning up
- remove references to animation and view
- removed anathema-string. it was an artifact from a previous version
- events
- raw mode and show cursor
- better value access api
- fixed a runtime bug where screen didn't seem to be cleared. removed unnecessary trait + impl
- adding meta data to the data context
- moved the default widgets into their own crate
- this has a lifetime error
- border layout
- implemented views and removed 'tpl lifetime
- updated README to use the new template syntax
- replaced unwrap with expect in the vm, as it's not a user error if the consts have changed!
- put test utils behind feature flag, enabled by default
- cleaned out TODOs
- cleanup TODOs
- renamed Store to Values and added get_ref<T>
- equality check for widgets
- change order of eq check to avoid unnecessary downcast_ref
- fixed bug with Border sides, added Frame to runtime
- frame
- bit of tidy up
- remove unsafe code and use constants instead of strings
- made ctrl-c it's own event
- mouse enabled on runtime
- wip
- wip
- wip
- wip
- hashmap instead
- benches
- use integer_hasher for path ids, criterion benchmarks
- addeding a new value type
- wip
- subscopes
- fixed map and list inserts / get
- scopes takes optional parent
- added truthynes check
- getting somewhere!
- generator
- about to burn it all down...
- paths transfer and widget meta
- getting closer
- change banana
- value sub collection
- untested node id
- wip
- wip
- wip
- wip
- wip
- wip
- wip
- wip
- before I break eeeeverything£
- all is broken
- I have to go to Sunday dinner now.
- wip
- wip
- fixed values
- fixed values
- all is broken, I blame caleb!
- moooaaar traaaits
- I really should be sleeping...
- One step closer
- Fix one, break another
- it's all compiling but it's super broken still
- code goes weeeeee
- happy birdah caleeeeb
- tests!
- broken broken broken
- omg it's almost funny how broken everything is
- oh my the rabbit hole...
- need to scope them values
- is that a horse in the distance?
- everything is about to get a bit more complicated, but food was good
- scoping values for loops, but it's not pretty
- why is my internet connection being all wonky today?
- everything is delightfully borked at this point, including my brain, so bed it is
- fixed all the bugs, except the ones that werent fixed
- oh look what a fine commit message this is!
- fixed test
- happy Sunday
- wip
- wip
- wip
- wip
- oh nodeid, such a fickle thing
- wip
- wip
- wip
- wip
- wip
- wip
- late night wip
- early morning wip
- some list stuff works
- integers as strings
- fixed a bug, probably made some more
- I wish I had a cup of tea right about now
- from 800 to 80 in 80 days?
- from 800 to 80 in 80 days?
- can add things? yes we can!
- caching? sounds like a fun idea...
- no way, it's working? well kind of
- caching? it's not a problem!
- we are off to grandmas house!
- Scoping is bad, this needs to be redone
- fixed scoping
- family time
- skill issue
- fixing tests?
- fixing tests?
- it's more test fixings
- remove the need to pass an attribute name to the attribtue parser
- renamed attribute parser to value parser
- let's break everything and then get in the car
- I'm so tired I just wanna sleep but at least the code compiles. It's not doing anything useful though
- this is deffs going to break a lot
- fixing tests
- let's make a mess
- whoops
- I'm going to break a lot now
- off we go
- omg it compiles!
- time to celebrate with a shake!
- time to celebrate that all tests (that I care about today) are passing
- some num stuff
- I don't remember, I'm sure it's something good
- so many new tests
- we got like, if statements, yeah!
- this is compiling, it might look like it's not, but it is
- now it's compiling for realisies
- cleanup tests
- this is not ruby
- lol broken
- either scoping works or eval of nodes work, but not both
- who remembers? I don't
- some tests are passing, yay?
- about to make life a lot more complicated
- I am commitet to spelllling
- food time? yes I think so
- finally, maps are working!
- bye oldie
- getting there..
- more listy things
- no main
- wip
- done with you for the day, time to move on to other things. See you tomorrow, code
- about time to clean up warnings
- trouble afoot
- wip
- off to grandmas house
- almost something but mostly nothing
- notey
- time to break everything
- so closure it is
- forgot the rest
- opsie
- bye bye benchie
- lol
- cleanup in isle warnings
- making some progress!
- kind of have control flow now!
- shuffling the ole runtime back in
- a bit of test fixery
- one step closer
- baby steps
- removed Value
- and or
- one step closer
- mooodulo
- started maps
- intentionally breaking
- let's start consolidating some of these structs
- factory context makes more sense
- data missing from a string still renders what's there
- about to wreck it
- cleanup
- remove scopevalue
- why
- about to make life harder
- let's try not to turn back from here, seems like the right path
- fixing tests
- deferred state value, but it's a bit messy right now
- about to break things again
- intentionally broken
- oh my breaking breaking
- meeh
- oh my lifetimes
- move widget container
- local scope
- one step closer
- remove old scope
- it's a bit of a mess but it's close
- omg 0 to 1
- can add but not quite remove
- more breaking
- strings be stringing again
- meh
- sleepy time
- value resolution
- breaking more things
- finally some more progress
- fixed values
- panic!
- padding
- styles...
- text has styles? yay
- yay
- cleanup
- cleanup
- cleanup
- testing
- alignment
- border
- there is a bug with the size of the border being calculated
- borders are broken, size wise
- border bug is not as bugged
- more tests!
- I'm a go and stream now
- allow maps in value expressions, though we don't do anything with them yet
- nesting the nestings
- nested state collections!
- insert is broken, but heading in a direction
- fixing some tests
- value insertion now works!
- going for it
- no more options
- proc macros, thanks to ModProcMacro
- manyhow
